manager/manager.go

```diff
@@ -0,0 +1,359 @@
+	backendOps := make(chan backendOp)
+	m.backendOps = backendOps
+
+	// This goroutine here serializes work so that we don't have concurrent commits or unwatches / updates / etc.
+	go func() {
+
+		for {
+			select {
+
+			case op := <-backendOps:
+				log.Debugln("Backend operation:", op)
+				if m.isLeader {
+					op.err <- op.f()
+				} else {
+					op.err <- fmt.Errorf("not-a-leader")
+				}
+
+			case <-stopWorkQueue:
+
+				log.Infoln("Stopping work queue.")
+				close(m.running)
+				log.Infoln("Manager stopped.")
+				return
+
+			case leader := <-notify:
+
+				// This channel has data only when there's been a leadership change.
+
+				log.Debugln("leader:", leader)
+				if leader {
+					m.onAssumeLeadership()
+				} else {
+					m.onLostLeadership()
+				}
+			}
+
+		}
+	}()
+
+	// Goroutine for handling all inbound leadership and control events.
+	go func() {
+
+		for {
+			select {
+
+			case <-m.stop:
+
+				log.Infoln("Stopping..")
+				close(stopWorkQueue)
+				close(notify)
+				return
+
+			case evt := <-leaderChan:
+				// This here handles possible duplicated events about leadership and fires only when there
+				// is a change.
+
+				m.lock.Lock()
+
+				current := m.isLeader
+
+				if evt.Status == leader.Unknown {
+					log.Warningln("Leadership status is uncertain:", evt.Error)
+
+					// if we are currently the leader then there's a possibility of split brain depending on
+					// the robustness of the leader election process.
+					// It's better to be conservative and not assume we can still be a leader...  just downgrade
+					// because the worst case is we stopped watching (and not have two masters running wild).
+
+					if m.isLeader {
+						m.isLeader = false
+					}
+
+				} else {
+					m.isLeader = evt.Status == leader.Leader
+				}
+				next := m.isLeader
+
+				m.lock.Unlock()
+
+				if current != next {
+					notify <- next
+				}
+
+			}
+		}
+
+	}()
+
+	return m.running, nil
+}
+
+// Stop stops the manager
+func (m *manager) Stop() {
+	if m.stop == nil {
+		return
+	}
+	m.leader.Stop()
+	close(m.stop)
+}
+
+func (m *manager) onAssumeLeadership() error {
+	log.Infoln("Assuming leadership")
+
+	// load the config
+	config := GlobalSpec{}
+
+	// load the latest version -- assumption here is that it's been persisted already.
+	err := m.snapshot.Load(&config)
+	if err != nil {
+		log.Warningln("Error loading config", err)
+		return err
+	}
+
+	log.Infoln("Loaded snapshot. err=", err)
+	if err != nil {
+		return err
+	}
+
+	m.lock.Lock()
+	m.currentConfig = config // make a copy
+	m.lock.Unlock()
+
+	return m.doWatchGroups(config)
+}
+
+func (m *manager) onLostLeadership() error {
+	log.Infoln("Lost leadership")
+
+	// Unwatch uses a cached version in memory.  This means that we are never able to
+	// unwatch without first watching.  If the manager crashes and comes back up again,
+	// there will be no possibility to unwatch existing groups.
+	m.lock.Lock()
+	config := m.currentConfig
```

Comment:
I agree that querying is better than an in-memory copy since I noted elsewhere in a comment that using the in-memory copy to notify the downstream plugins won't work if the manager exited and comes back up.  So I am changing to querying *for now* the proxied group plugin.   Note that this isn't the general case where a global config has multiple groups with different group plugins implementations running (instead of just the single one proxied here).   To construct the global spec requires querying *all* the group plugins (imagine one that's group-default and one that's backed by some autoscaling group) to construct this global collection of groups and their specs.

Anyway, added a TODO / comment and changing to querying the backend plugin for now.